packages/front-end/components/Features/FeatureModal.tsx

```diff
@@ -61,62 +62,88 @@ function parseDefaultValue(
   }
 }
 
+const initEnvironmentSettings = ({
+  environments,
+  featureToDuplicate,
+  permissions,
+  project,
+}: {
+  environments: ReturnType<typeof useEnvironments>;
+  featureToDuplicate?: FeatureInterface;
+  permissions: ReturnType<typeof usePermissions>;
+  project: string;
+}): Record<string, FeatureEnvironment> => {
+  const envSettings: Record<string, FeatureEnvironment> = {};
+
+  environments.forEach((e) => {
+    const canPublish = permissions.check("publishFeatures", project, [e.id]);
+    const defaultEnabled = canPublish ? e.defaultState ?? true : false;
+    const enabled = canPublish
+      ? featureToDuplicate?.environmentSettings?.[e.id]?.enabled ??
+        defaultEnabled
+      : false;
+    const rules = featureToDuplicate?.environmentSettings?.[e.id]?.rules ?? [];
+
+    envSettings[e.id] = { enabled, rules };
+  });
+
+  return envSettings;
+};
+
 export default function FeatureModal({
   close,
   onSuccess,
   inline,
   cta = "Create",
   secondaryCTA,
+  featureToDuplicate,
 }: Props) {
   const { project, refreshTags } = useDefinitions();
   const environments = useEnvironments();
   const permissions = usePermissions();
-
   const { refreshWatching } = useWatching();
 
-  const defaultEnvSettings: Record<string, FeatureEnvironment> = {};
-  environments.forEach((e) => {
-    let enabled = e.defaultState ?? true;
-    if (!permissions.check("publishFeatures", project, [e.id])) enabled = false;
-
-    defaultEnvSettings[e.id] = {
-      enabled,
-      rules: [],
-    };
+  const initialEnvSettings = initEnvironmentSettings({
+    environments,
+    featureToDuplicate,
+    permissions,
+    project,
   });
 
   const form = useForm({
     defaultValues: {
-      valueType: "boolean",
-      defaultValue: getDefaultValue("boolean"),
+      valueType: featureToDuplicate ? featureToDuplicate.valueType : "boolean",
```

Comment:
This might be a bit less cumbersome if you use the format:
`valueType: featureToDuplicate?.valueType ?? "boolean"`

Or even consider abstracting the duplicate feature overrides away.

Reply:
One way I did this for the Add/Edit event webhook functionality (still WIP 🚧) is I added a `mode` prop that took a mode `edit` | `add` (yours could be `duplicate`) and prefilled form data that only shows up in `edit` mode.

- Here's the definition of the mode type: https://github.com/growthbook/growthbook/blob/main/packages/front-end/components/EventWebHooksPage/utils.ts#L27-L32
- Here's the implementation: https://github.com/growthbook/growthbook/blob/main/packages/front-end/components/EventWebHooksPage/EventWebHookAddEditModal/EventWebHookAddEditModal.tsx#L25-L34

I think this works better for larger forms and reads a bit easier because you can know the full default state of the form by scanning and/or checking a single value (the `mode` and its accompanying form state) rather than having to mentally parse all each individual property, and if you have 20 properties and they're all using ternaries or null coalescing (I don't personally have a preference in this case since we aren't working with numbers), that could take a bit longer to understand. In my example, `mode.data` is always present if the `mode.mode === 'edit'` and never prsent if `mode.mode === 'add'`. I acknowledge this probably needs better naming, but that's what I went with.

Reply:
Yes - refactored to a single abstraction method ... `genDefaultFormValues`